File: backend/internal/blockparty/engine/move_test.go
package engine

import "testing"

func st(b Board) State {
	return State{Board: b, Score: 0, Rng: 555, RngCallCount: 0, Modifier: "standard", Moves: 0, Over: false}
}

func TestApplyMove_LeftMerge(t *testing.T) {
	s := st(Board{2, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0})
	ns, changed := ApplyMove(s, "L")
	if !changed {
		t.Fatal("expected changed")
	}
	if ns.Board[0] != 4 {
		t.Fatalf("board[0]=%d want 4", ns.Board[0])
	}
	if ns.Score != 4 {
		t.Fatalf("score=%d want 4", ns.Score)
	}
	if ns.RngCallCount != 2 {
		t.Fatalf("rngCallCount=%d want 2", ns.RngCallCount)
	}
}

func TestApplyMove_Right(t *testing.T) {
	s := st(Board{2, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0})
	ns, _ := ApplyMove(s, "R")
	if ns.Board[3] != 4 {
		t.Fatalf("board[3]=%d want 4", ns.Board[3])
	}
}

func TestApplyMove_Up(t *testing.T) {
	s := st(Board{0, 0, 0, 0, 2, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0})
	ns, _ := ApplyMove(s, "U")
	if ns.Board[0] != 4 {
		t.Fatalf("board[0]=%d want 4", ns.Board[0])
	}
}

func TestApplyMove_NoOp(t *testing.T) {
	s := st(Board{2, 4, 8, 16, 4, 8, 16, 32, 8, 16, 32, 64, 16, 32, 64, 128})
	ns, changed := ApplyMove(s, "L")
	if changed {
		t.Fatal("expected no-op")
	}
	if ns.RngCallCount != 0 {
		t.Fatalf("rngCallCount=%d want 0", ns.RngCallCount)
	}
}

File: frontend/src/game/engine/move.test.ts
import { describe, it, expect } from "vitest";
import { applyMove, LINE_INDICES } from "./move";
import type { GameState } from "./types";

function stateWith(board: number[]): GameState {
  return { board, score: 0, rng: 555, rngCallCount: 0, modifier: "standard", moves: 0, over: false };
}

describe("LINE_INDICES", () => {
  it("has 4 lines of 4 indices for each direction", () => {
    for (const dir of ["U", "R", "D", "L"] as const) {
      expect(LINE_INDICES[dir].length).toBe(4);
      for (const line of LINE_INDICES[dir]) expect(line.length).toBe(4);
    }
  });
});

describe("applyMove", () => {
  it("moves left and merges a row", () => {
    // row 0 = [2,2,0,0]; expect it to become [4,0,0,0] before spawn
    const s = stateWith([2, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]);
    const next = applyMove(s, "L");
    expect(next.board[0]).toBe(4);
    expect(next.score).toBe(4);
    expect(next.moves).toBe(1);
    // a spawn happened => exactly one new tile appeared somewhere in indices 1..15
    expect(next.rngCallCount).toBe(2);
    expect(next.board.filter((v) => v !== 0).length).toBe(2);
  });

  it("moves right (tiles pull toward the high index of each row)", () => {
    const s = stateWith([2, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]);
    const next = applyMove(s, "R");
    expect(next.board[3]).toBe(4);
  });

  it("moves up (tiles pull toward row 0 of each column)", () => {
    // column 0 = indices 0,4,8,12 = [0,2,2,0]; up => [4,...]
    const s = stateWith([0, 0, 0, 0, 2, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0]);
    const next = applyMove(s, "U");
    expect(next.board[0]).toBe(4);
  });

  it("returns the SAME state (no spawn, no rng) on a no-op move", () => {
    // full-left-packed row cannot move further left; whole board is a no-op for L
    const s = stateWith([2, 4, 8, 16, 4, 8, 16, 32, 8, 16, 32, 64, 16, 32, 64, 128]);
    const next = applyMove(s, "L");
    expect(next).toBe(s); // identity — unchanged
    expect(next.rngCallCount).toBe(0);
  });
});
